pageserver/src/tenant/storage_layer.rs

```diff
@@ -255,24 +300,44 @@ impl LayerAccessStats {
         ret
     }
 
-    fn most_recent_access_or_residence_event(
-        &self,
-    ) -> Either<LayerAccessStatFullDetails, LayerResidenceEvent> {
+    /// Get the latest access timestamp, falling back to latest residence event, falling back to `fallback`.
+    ///
+    /// The `fallback` must be supplied for the case where there has not yet been a call to the
+    /// [`record_residence_event`] method.
+    /// If `fallback` needs to be used, we log a rate-limited warning in global scope.
+    pub(crate) fn latest_activity<F>(&self, fallback: F) -> SystemTime
+    where
+        F: FnOnce() -> SystemTime,
+    {
         let locked = self.0.lock().unwrap();
         let inner = &locked.for_eviction_policy;
         match inner.last_accesses.recent() {
-            Some(a) => Either::Left(*a),
+            Some(a) => a.when,
             None => match inner.last_residence_changes.recent() {
-                Some(e) => Either::Right(e.clone()),
-                None => unreachable!("constructors for LayerAccessStats ensure that there's always a residence change event"),
-            }
-        }
-    }
-
-    pub(crate) fn latest_activity(&self) -> SystemTime {
-        match self.most_recent_access_or_residence_event() {
-            Either::Left(mra) => mra.when,
-            Either::Right(re) => re.timestamp,
+                Some(e) => e.timestamp,
+                None => {
+                    // TODO: use type system to avoid the need for `fallback`.
+                    // The approach in https://github.com/neondatabase/neon/pull/3775
+                    // could be used to enforce that a residence event is recorded
+                    // before a layer is added to the layer map. We could also have
+                    // a layer wrapper type that holds the LayerAccessStats, and ensure
+                    // that that type can only be produced by inserting into the layer map.
+                    static WARN_RATE_LIMIT: Lazy<Mutex<(usize, RateLimitClosure)>> =
+                        Lazy::new(|| {
+                            Mutex::new((0, RateLimitClosure::new(Duration::from_secs(10))))
+                        });
+                    let mut guard = WARN_RATE_LIMIT.lock().unwrap();
+                    guard.0 += 1;
+                    let occurences = guard.0;
+                    guard.1.call(move || {
+                        // Spawn a new task to have a new tracing root. It's ok, it only happens every 10 secs.
+                        tokio::spawn(async move {
+                            warn!(occurences, "latest_activity not available, this is an implementation bug, using fallback value");
+                        });
```

Comment:
What is the point of spawning here compared to just doing it within the closure?

Reply:
This question still holds.

Reply:
```suggestion
                        warn!(parent: None, occurences, "latest_activity not available, this is an implementation bug, using fallback value");
```

This should do the trick, if it really was to be detached, but I don't understand what is the rationale? We can query logs with regex.

Reply:
Sorry, I thought I had replied here.

Yeah, the point of the spawn was to log in global scope.

I always forget about the `parent: None`, will use that.